### ISSUE TITLE
Bug: Playback seeks backwards when speed is increased, and seeks forwards when speed is decreased

### ISSUE THREAD

**Issue by mesvam** [mpv-player/mpv#13513]
### Important Information

Provide following Information:
- mpv version: 
```
> mpv --version
mpv v0.37.0-337-gbd5b80ba Copyright © 2000-2024 mpv/MPlayer/mplayer2 projects
 built on Feb 18 2024 00:09:08
libplacebo version: v6.338.0-77-g3ba18d5-dirty
FFmpeg version: N-113670-g0895ef0d6
FFmpeg library versions:
   libavutil       58.39.100
   libavcodec      60.39.101
   libavformat     60.21.100
   libswscale      7.6.100
   libavfilter     9.17.100
   libswresample   4.13.100
```
- Platform and Version: Windows 11
- Source of the mpv binary: https://sourceforge.net/projects/mpv-player-windows/files/64bit-v3/mpv-x86_64-v3-20240218-git-bd5b80b.7z/download
- If known which version of mpv introduced the problem: unknown
- GPU model, driver and version: GTX 1080
- Possible screenshot or video of visual glitches: none

### Reproduction steps

- Start with clean config/profile
- add this debug script to the profile under the scripts directory
```
require 'mp.msg'
mp.set_property("msg-level", string.format("%s=debug", mp.get_script_name()))

local function s1()
    local before = mp.get_property_native('audio-pts')
    mp.msg.debug(string.format("before: %f", before))
    mp.set_property_native("speed", 1)
    local after = mp.get_property_native('audio-pts')
    mp.msg.debug(string.format("after : %f", after))
    mp.msg.debug(string.format("diff  : %f", after-before))
end
local function s2()
    local before = mp.get_property_native('audio-pts')
    mp.msg.debug(string.format("before: %f", before))
    mp.set_property_native("speed", 4)
    local after = mp.get_property_native('audio-pts')
    mp.msg.debug(string.format("after : %f", after))
    mp.msg.debug(string.format("diff  : %f", after-before))
end

mp.add_key_binding("1", "slow", s1)
mp.add_key_binding("2", "fast", s2)
```
- play a video or audio file from the console (e.g. `mpv.com video.mkv`) to show debug messages
- toggle the speed with the keys `1` and `2` during playback. `1` sets speed to 1x and `2` sets speed to 4x
- If necessary, use screen capture software or a phone to record the screen to capture video seeking behavior

### Expected behavior

Playback time should not change (no seeking). `audio-pts` before and after the speed change should be identical or very close. Ideally the console should show something like
```
[debug] before: 12.071627
[debug] after : 12.071627
[debug] diff  : 0
```

### Actual behavior

The player actually seems to seek backwards and forwards by a set duration when the speed changes, as seen on the OSD. There is a large difference in `audio-pts` before and after the speed change.
```
[debug] before: 12.071627
[debug] after : 11.263086
[debug] diff  : -0.808541
```
The time difference increases as the speed difference increases. In severe cases, when the speed changes a lot (e.g. from 1x to 16x), `audio-pts` can actually be negative, seeking beyond the boundaries of the file. Here is the output for a 1x to 16x speed change.
```
[debug] before: 1.047906
[debug] after : -3.047085
[debug] diff  : -4.094991
```
mpv seeks the file to -3.047085 and the difference is over 4 seconds!

Why this matters: I have scripts that depend on always having accurate playback time

### Log file

[mpv.log](https://github.com/mpv-player/mpv/files/14326847/mpv.log)

### Sample files

Any video or audio file will work, but if you use this diagnostic video https://www.youtube.com/watch?v=szoOsG9137U (download with `yt-dlp`), you can clearly see the video seeking on a screen recording

**Comment by mesvam** [comment]
Additionally, mpv does not seem to realize that it is seeking; if you add an event hook to `seek`, it never fires
```lua
mp.register_event("seek", function(a)
		mp.msg.debug("seek")
	end
)
```

**Comment by Dudemanguy** [comment]
mpv isn't seeking. What's happening is that mpv adjusts the returned audio-pts value based on speed. If you pause the video, you can get different different values simply by changing the playback speed. I agree that it doesn't really make sense.

**Comment by mesvam** [comment]
It's not just that the returned value is wrong, but both the video and the audio position are changed. If you play that diagnostic video for example, the I can see the frame numbers repeat during speed is increased, so it's not purely an audio issue, but the effect on audio is identical and it can be seen on an audio spectrogram as well.

Maybe mpv know it's seeking, but I don't know how you can call that anything else but seeking. I'll try to make a screen recording as an example/proof later.

**Comment by Dudemanguy** [comment]
From the linked sample, I certainly am not able to see frame numbers repeating or anything of that nature when changing speed. What are you doing exactly to see this? I fully understand why the audio-pts value was weird (hence the linked PR), but I don't see how pseudo-seeking could be happening.

**Comment by mesvam** [comment]
Here's a screen recording of the video seeking issue. (Note: video frame rate is 29.97fps, playback frame rate is 60fps) When you analyze it frame by frame, at speed = 1 each frame should be displayed twice, and at speed = 4, it should skip every other frame.

https://github.com/mpv-player/mpv/assets/14833020/cebd9a4d-2d87-43ce-95aa-9f56e75533f7

But actual behavior is that when switching speed from 4 to 1, it skips 25 frames (0.833s), from frame 261 to 287. Even weirder, when switching speed from 1 to 4, it gets stuck on frame 68 for 12 frames (0.4s), then plays at speed 2 until frame 77 (displays every frame rather than every other frame).

I'm still trying to find a minimum reproducible example for the audio portion, but here's what the behavior looks like on the audio spectrogram from a more complex script when playing back a frequency sweep.
![image](https://github.com/mpv-player/mpv/assets/14833020/6a6d5fbd-4f89-4706-84e9-ef6595b31e00)
The speed transitions should look like a sudden change in the slope, but there is also a discontinuity, indicating skipped audio.

**Comment by Dudemanguy** [comment]
OK, indeed I see what you mean. Something weird is going on with the filters.

**Comment by mesvam** [comment]
When speed = 1, scaletempo2 filter is removed, and it's added when speed != 1. So my theory right now is that the filter addition/removal causes seeking-like behavior since it resets the audio buffers somehow.

**Comment by Dudemanguy** [comment]
Can you test this hack patch? I think I understand the problem but I'm not sure what the correct solution is yet.

```
diff --git a/player/command.c b/player/command.c
index e81874010b..086bbea51a 100644
--- a/player/command.c
+++ b/player/command.c
@@ -7138,6 +7138,7 @@ void mp_option_change_callback(void *ctx, struct m_config_option *co, int flags,
 
     if (opt_ptr == &opts->playback_speed) {
         update_playback_speed(mpctx);
+        mpctx->update_speed_wait = mp_time_ns() + MP_TIME_S_TO_NS(1);
         mp_wakeup_core(mpctx);
     }
 
diff --git a/player/core.h b/player/core.h
index d995a88366..36f182dcd0 100644
--- a/player/core.h
+++ b/player/core.h
@@ -333,6 +333,7 @@ typedef struct MPContext {
     // Redundant values set from opts->playback_speed and speed_factor_*.
     // update_playback_speed() updates them from the other fields.
     double audio_speed, video_speed;
+    double update_speed_wait;
     bool display_sync_active;
     int display_sync_drift_dir;
     // Timing error (in seconds) due to rounding on vsync boundaries
diff --git a/player/video.c b/player/video.c
index 70ef18958b..e2d213b062 100644
--- a/player/video.c
+++ b/player/video.c
@@ -592,7 +592,9 @@ static void update_avsync_before_frame(struct MPContext *mpctx)
 
     if (mpctx->video_status < STATUS_READY) {
         mpctx->time_frame = 0;
-    } else if (mpctx->display_sync_active || vo->opts->video_sync == VS_NONE) {
+    } else if (mpctx->display_sync_active || vo->opts->video_sync == VS_NONE ||
+               mp_time_ns() < mpctx->update_speed_wait)
+    {
         // don't touch the timing
     } else if (mpctx->audio_status == STATUS_PLAYING &&
                mpctx->video_status == STATUS_PLAYING &&
@@ -831,6 +833,9 @@ static void handle_display_sync_frame(struct MPContext *mpctx,
     if (resample && using_spdif_passthrough(mpctx))
         return;
 
+    if (mp_time_ns() < mpctx->update_speed_wait)
+        return;
+
     double vsync = vo_get_vsync_interval(vo) / 1e9;
     if (vsync <= 0)
         return;
```

**Comment by mesvam** [comment]
I don't think I have time to setup a build environment right now, but I could test a binary if you have one.

**Comment by Dudemanguy** [comment]
Ah, you can try the windows builds linked [https://github.com/mpv-player/mpv/pull/13550#issuecomment-1962715684](here). I added that commit to that branch.

**Comment by mesvam** [comment]
Thanks,

The audio-pts seems fixed. I also seem unable to induce audio skipping, but since I don't have a reliable reproduction, I can't be sure that is also fixed.

But video frame rate change is still delayed a few frames, and there is still a brief transition period where speed = 2.

**Comment by Dudemanguy** [comment]
I believe there are essentially two problems here.

1. Inserting the audio filter has a certain amount of delay before it actually kicks in. I'm not sure what could be done about that fact. Adding any type of audio filter seems to take some time and is not instant.
2. Because of that delay in adding the audio filter, audio/vsync is completely wrong and causes the "seek-like" behavior that you mentioned. What's happening is that frame times are getting adjusted based on calculated delay values but in that time period what is calculated is essentially totally bogus. So the frame times get adjusted to weird places and this is why you see frame skipping and things like that.

While it would be nice if the audio filter could act instantly, I'm not sure how realistic that is and I also do not think that in and of itself is a pressing problem. The second one is definitely quite bad though. The approach with the patch is to simply stop a/v adjusting while the filter is being setup which is what `--video-sync=desync` or `--video-sync=display-desync` do. The reason it's a hack is because it's a completely arbitrary 1 second wait. I'm not sure yet how to properly know when the filter is set in place and actually active.

**Comment by Dudemanguy** [comment]
Actually, the audio filter part was a red herring. There's a delay even with `--no-audio-pitch-correction` which inserts no filter at all. It's just from mpv itself. I think I'll have a reasonable enough way to handle this in a bit.

**Comment by Dudemanguy** [comment]
>There is still swresample filter changing the sample rate in this case.

Ah yes. More difficult to verify if this causes a delay or not. I guess it might due to audio draining.

@mesvam: Could you give the build in https://github.com/mpv-player/mpv/pull/13550 a shot again? I changed the second commit.

**Comment by mesvam** [comment]
The audio component is actually a bigger issue, at least for me. I am working on a script that changes playback speed based on volume changes and other events, which means speed will change quite often, maybe up to 5 times in 10 seconds. I can somewhat compensate for things like the 200ms audio buffer if the delay is a fixed duration, but if the audio filter delay varies (such as based on speed) then it becomes much more complicated.

Also, I find the video lagging/skipping less noticeable, especially at faster speed, since it's basically a cut at the wrong time. But skipping even 300ms of audio means you can miss entire spoken words or musical notes, which can sound quite odd.

**Comment by mesvam** [comment]
Is there a way to force an audio buffer duration? The documentation says `--audio-buffer` only suggests a maximum duration, but mpv is still in control. I think it would help me debug the issue further if there was a way to control the audio-buffer

**Comment by mesvam** [comment]
> Could you give the build in https://github.com/mpv-player/mpv/pull/13550 a shot again? I changed the second commit.

I only checked the video frame skipping. Speed 1 -> 4 transition is totally fixed. Speed 4 -> 1 transition still has a few frames where it's playing at speed = 2

**Comment by mesvam** [comment]
> If you could verify the audio portion of this, that would be great. It should hopefully work, but I have not done any fancy testing or anything on it.

audio-pts is now nil, so I can't check that (I'm assuming that's just because it's unimplemented in the test build?) I can no longer induce audio skipping either.

In any case, even if this doesn't solve all the issues, still seems like an improvement.

**Comment by Dudemanguy** [comment]
Updated it again. `audio-pts` should be non-nil now.

**Comment by richardpl** [comment]
This is by mpv design, it is resetting filter state (for both libavfilter and mpv internal filters) each time seek happens or any changes to filter options that reset filter state (the only current approach to not reset filters is via commands for libavfilter filter code and that one still will not help when seeking). That is one of reasons why actually gap-less audio playback is not possible with audio filters active (without clicks and similar).

Fixing this would be major mpv rewrite that I have not yet attempted.

**Comment by richardpl** [comment]
Any filter qualify, even internal mpv filters, like rubberband, and two scaletempo filters, when seeking it will reset filter state (which might be or not be desirable), also it will reset filter state on each new input switch after previous input 'end of file' - this will always break gap-less playback when non-trivial filtering is used. This is just one design issue that I wanted to raise and is marginally related with issue described in this report.

**Comment by mesvam** [comment]
It appears this change has affected some other seeking & speed change behavior, for which I have filed another bug here: https://github.com/mpv-player/mpv/issues/13613

**Comment by mesvam** [comment]
Since 9ac791c reverts the flawed fix e3af545, can we reopen this issue?
